secrets/certregistry/certregistry.go

```diff
@@ -0,0 +1,102 @@
+package certregistry
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// CertRegistry object holds TLS certificates to be used to terminate TLS connections
+// ensuring syncronized access to them.
+type CertRegistry struct {
+	lookup map[string]*tls.Certificate
+	mx     *sync.Mutex
+}
+
+// NewCertRegistry initializes the certificate registry.
+func NewCertRegistry() *CertRegistry {
+	l := make(map[string]*tls.Certificate)
+
+	return &CertRegistry{
+		lookup: l,
+		mx:     &sync.Mutex{},
+	}
+}
+
+func (r *CertRegistry) getCertByKey(key string) (*tls.Certificate, bool) {
+	cert, found := r.lookup[key]
+	if !found {
+		log.Debugf("certificate not found in registry - %s", key)
+		return nil, false
+	}
+
+	return cert, true
+}
+
+func (r *CertRegistry) addCertToRegistry(key string, cert *tls.Certificate) error {
+	r.lookup[key] = cert
+
+	return nil
+}
+
+// SyncCert takes a host and TLS certificate and saves them to the registry with the
+// host as the key. If the cert already exists it will be updated or added otherwise.
+func (r *CertRegistry) SyncCert(host string, cert *tls.Certificate) {
```

Comment:
I think this method should return error instead of logging - this way we let caller to handle the error (i.e. log it).

Reply:
I would also rename the method and update the doc comment
```go
// Configures certificate for the host if no configuration exists or
// if certificate is valid (`NotBefore` field) after previously configured certificate.
func (r *CertRegistry) ConfigureCertificate(host string, cert *tls.Certificate) error
```